File: Java/Tree/_297_SerializeandDeserializeBinaryTree.java
import java.util.*;
public class _297_SerializeandDeserializeBinaryTree {
	// lavel order ?
	// Encodes a tree to a single string.
	
	// as "[1,2,3,null,null,4,5]"
	
	// Encodes a tree to a single string.
	public String serialize(TreeNode root) {
	    ArrayList<String> list = new ArrayList<>();
	    LinkedList<TreeNode> q = new LinkedList<>();
	    q.offer(root);
	 
	    while (!q.isEmpty()) {
	        TreeNode h = q.poll();
	        if (h == null) {
	            list.add("#");
	        } else {
	            list.add("" + h.val);
	            q.offer(h.left);
	            q.offer(h.right);
	        }
	    }
	 
	    return String.join(",", list);
	}
	 
	// Decodes your encoded data to tree.
	public TreeNode deserialize(String data) {
	    String[] arr = data.split(",");
	    if (arr[0].equals("#")) {
	        return null;
	    }
	 
	    TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
	    LinkedList<TreeNode> q = new LinkedList<>();
	    q.offer(root);
	 
	    int i = 1;
	 
	    while (!q.isEmpty()) {
	        TreeNode h = q.poll();
	        if (h != null) {
	            TreeNode left = null;
	            if (!arr[i].equals("#")) {
	                left = new TreeNode(Integer.parseInt(arr[i]));
	            }
	            h.left = left;
	            q.offer(left);
	            i++;
	 
	            TreeNode right = null;
	            if (!arr[i].equals("#")) {
	                right = new TreeNode(Integer.parseInt(arr[i]));
	            }
	            h.right = right;
	            q.offer(right);
	            i++;
	        }
	    }
	 
	    return root;
	}
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		TreeNode t = new TreeNode(1);
		t.left = new TreeNode(2);
		t.right = new TreeNode(3);
		t.right.left = new TreeNode(4);
		t.right.right = new TreeNode(5);
		_297_SerializeandDeserializeBinaryTree sdb = new _297_SerializeandDeserializeBinaryTree();
		String data = "1,2,3,#,#,4,5,#,#,#,#";
		System.out.println(sdb.serialize(t));
		System.out.println(sdb.deserialize(data));

	}

}

File: Python/Leetcode Daily Practice/Tree/297. Serialize and Deserialize Binary Tree.py
"""
297. Serialize and Deserialize Binary Tree
Hard

Share
Serialization is the process of converting a data structure or object into a sequence of 
bits so that it can be stored in a file or memory buffer, or transmitted across a network connection 
link to be reconstructed later in the same or another computer environment.

Design an algorithm to serialize and deserialize a binary tree. There is no restriction on how your 
serialization/deserialization algorithm should work. You just need to ensure that a binary tree can be 
serialized to a string and this string can be deserialized to the original tree structure.

Example: 

You may serialize the following tree:

    1
   / \
  2   3
     / \
    4   5

as "[1,2,3,null,null,4,5]"
1,2,None 
"""
class Codec:
    def serialize(self, root):
        """ 
        Encodes a tree to a single string.
        :type root: TreeNode
        :rtype: str
        
        You may serialize the following tree:

            1
           / \
          2   3
             / \
            4   5

        as "[1,2,3,null,null,4,5]"
        """
        def dfs(root):
            if not root:
                res.append("None,")
                return 
            res.append(str(root.val)+",")
            dfs(root.left)
            dfs(root.right)
            
        res = []
        dfs(root)
        return "".join(res)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        :type data: str
        :rtype: TreeNode
        """
        def rdeserialize(l):
            if l[0] == 'None':
                l.pop(0)
                return None
                
            root = TreeNode(l[0])
            l.pop(0)
            root.left = rdeserialize(l)
            root.right = rdeserialize(l)
            return root

        data_list = data.split(',')
        root = rdeserialize(data_list)
        return root